arkouda/dataframe.py

```diff
@@ -5326,6 +5609,322 @@ def from_return_msg(cls, rep_msg):
+            key = key.values
+        if is_supported_scalar(key) and self.df.index.dtype == dtype(type(key)):
+            return self.df._get_rows(indexof1d(array([key]), self.df.index.values))
+
+        if isinstance(key, pdarray) and key.dtype == self.df.index.dtype:
+            return self.df._get_rows(indexof1d(key, self.df.index.values))
+
+        if isinstance(key, slice):
+            if key.start is not None and not akany(in1d(array([key.start]), self.df.index.values)):
+                raise KeyError(f"Index {key.start} not found in DataFrame index")
+            if key.stop is not None and not akany(in1d(array([key.stop]), self.df.index.values)):
+                raise KeyError(f"Index {key.stop} not found in DataFrame index")
+
+            start_idx = (indexof1d(array([key.start]), self.df.index.values)[0]
+                         if key.start is not None else 0)
+            stop_idx = (indexof1d(array([key.stop]), self.df.index.values)[0] + 1
+                        if key.stop is not None else self.df.index.size)
+
+            indices = arange(start_idx, stop_idx)
+            return self.df._get_rows(indices)
+
+        if isinstance(key, pdarray) and key.dtype == akbool:
+            return self.df._get_rows(key)
+
+        return None
+
+    def _get_row_col(self, row_key, col_key):
+        return self[row_key][col_key]
+
+    def __setitem__(self, key, val):
+        if isinstance(key, tuple) and len(key) == 2:
+            self._set_row_col(key[0], key[1], val)
+            return
+        else:
+            raise ValueError(
+                "Invalid key type. '.loc' indexing only supports keys with row and column selectors."
+            )
+
+    def _set_row_col(self, row_key, col_key, val):
+        if isinstance(row_key, list):
+            row_key = array(row_key)
+        if isinstance(row_key, Series):
+            row_key = row_key.values
+        if is_supported_scalar(col_key) and col_key not in self.df.columns.values:
+            self.df._add_column(col_key, dtype(type(val)))
+
+        if is_supported_scalar(val):
+            return self._set_row_col_scalar_val(row_key, col_key, val)
+        else:
+            assert (isinstance(val, (pdarray, Series, Strings, SegArray))), "Invalid value type"
+            return self._set_row_col_vector_val(row_key, col_key, val)
+
+    def _set_row_col_scalar_val(self, row_key, col_key, val):
+        if is_supported_scalar(row_key):
+            if not self.df.index.dtype == dtype(type(row_key)):
+                raise TypeError("Row key must be of the same type as the DataFrame index")
+            if akany(in1d(array([row_key]), self.df.index.values, invert=True)):
+                self.df._add_new_rows(row_key)
+            # updating a single row
+            row_idx = indexof1d(array([row_key]), self.df.index.values)
+            if row_idx.size == 0:
+                raise ValueError(f"Index {row_key} not found in DataFrame index")
```

Comment:
also this logic depends on the size of the return being `0` when the key is not found. If my PR gets merged which updates `indexof1d` to use `find`, this will no longer be the case and it will return `-1` instead. so if excluding indices that are not found is the desired behavior, then I need to update https://github.com/Bears-R-Us/arkouda/pull/3169

```python
>>> ak.find(ak.array([7]), ak.arange(5))
array([-1])

>>> ak.indexof1d(ak.array([7]), ak.arange(5))
array([])
```

Reply:
That is the desired behavior! The return of indexof1d should just be a list of indices, potentially empty.

Reply:
sweet! I'll get that updated! I also want to note that will change how we want to handle these type of checks:
https://github.com/Bears-R-Us/arkouda/pull/3109#discussion_r1599199235

which should be small change so I can handle that as part of the indexof1d update!